packages/edit-site/src/components/page-patterns/pagination.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	useLayoutEffect,
+	useEffect,
+	useRef,
+	useState,
+	startTransition,
+} from '@wordpress/element';
+import { __experimentalHStack as HStack, Button } from '@wordpress/components';
+
+export default function PatternsPagination( {
+	patterns,
+	page,
+	setPage,
+	getTotalPages,
+} ) {
+	const [ totalPages, setTotalPages ] = useState( page );
+	const getTotalPagesRef = useRef( getTotalPages );
+	useLayoutEffect( () => {
+		getTotalPagesRef.current = getTotalPages;
+	} );
+
+	// Refetch total pages when `patterns` changes.
+	// This is not a good indicator of when to refetch the total pages,
+	// but the only one we have for now.
+	useEffect( () => {
```

Comment:
This is not fine-tuned. It currently will fire a lot of requests since we're not sure when to fetch this info. A more elegant solution will be baking this into `getEntityRecords` somehow. The REST API already supports this by returning the `X-Wp-Totalpages` response header. However, `getEntityRecords` (and `apiFetch` under the hood) doesn't support it yet. It might require multiple levels of coordination across different packages to implement this though.